test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/MetricsApmIT.java

```diff
@@ -91,21 +88,35 @@ public void testApmIntegration() throws Exception {
                 var samples = (Map<String, Object>) metricset.get("samples");
 
                 samples.forEach((key, value) -> {
-                    var assertion = sampleAssertions.get(key);// sample name
-                    if (assertion != null) {
-                        logger.info("Matched {}", key);
+                    var valueAssertion = valueAssertions.get(key);// sample name
+                    if (valueAssertion != null) {
+                        logger.info("Matched {}:{}", key, value);
                         var sampleObject = (Map<String, Object>) value;
-                        if (assertion.test(sampleObject)) {// sample object
+                        if (valueAssertion.test(sampleObject)) {// sample object
+                            logger.info("{} assertion PASSED", key);
+                            valueAssertions.remove(key);
+                        } else {
+                            logger.error("{} assertion FAILED", key);
+                        }
+                    }
+                    var histogramAssertion = histogramAssertions.get(key);
+                    if (histogramAssertion != null) {
+                        logger.info("Matched {}:{}", key, value);
+                        var samplesObject = (Map<String, Object>) value;
+                        var counts = ((Collection<Integer>) samplesObject.get("counts")).stream().mapToInt(Integer::intValue).sum();
+                        var remaining = histogramAssertion - counts;
+                        if (remaining == 0) {
                             logger.info("{} assertion PASSED", key);
-                            sampleAssertions.remove(key);
+                            histogramAssertions.remove(key);
                         } else {
-                            logger.error("{} assertion FAILED: {}", key, sampleObject.get("value"));
+                            logger.info("{} assertion PENDING: {} remaining", key, remaining);
+                            histogramAssertions.put(key, remaining);
```

Comment:
If I understand these changes, I think what's happening is that each key could have multiple samples, and together they should add up to the expected quantity?

Have you considered scanning the samples to aggregate a summary of the _actual_ results, and then assert that this equals the _expected_ results? That is a more typical pattern for unit tests, and might be easier for others to follow, rather than a loop that interleaves mutating data structures with assertions.

Reply:
That's true, it would be nicer, but has the downside of being more time consuming. 

Today we are reading samples and counting down till zero, so as soon as we have enough data (can be all in the first message), we good. Otherwise we keep listening for more data and check again until TIMEOUT.
If we aggregate and then check against expected, we will have first to wait until TIMEOUT, then check (sure we can have an assertBusy, but that is more or less the same, only with multiple timeouts).

In the worst scenario this is the same, but the test usually is quicker and pass as soon as we meet the criteria.

Reply:
That said, I did a minimal change to fix the test, and I have to say I also find the mutating data structures with assertions pattern very very strange.
Probably it would be better to rewrite this test in a way similar to what you suggest (with assertBusy), but I would leave it for a followup PR.

Reply:
Makes sense!